src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                    item->_instruction = isConflict ? CSYNC_INSTRUCTION_CONFLICT : CSYNC_INSTRUCTION_UPDATE_METADATA;
+                    item->_direction = isConflict ? SyncFileItem::None : SyncFileItem::Down;
+                }
+            }
+            if (dbEntry._type == ItemTypeVirtualFile)
+                item->_type = ItemTypeVirtualFileDownload;
+            if (item->_file.endsWith(_discoveryData->_syncOptions._virtualFileSuffix)) {
+                item->_file.chop(_discoveryData->_syncOptions._virtualFileSuffix.size());
+                item->_type = ItemTypeVirtualFileDownload;
+            }
+            item->_previousSize = localEntry.size;
+            item->_previousModtime = localEntry.modtime;
+        } else if (typeChange) {
+            item->_instruction = CSYNC_INSTRUCTION_TYPE_CHANGE;
+            item->_direction = SyncFileItem::Up;
+            item->_checksumHeader.clear();
+            item->_size = localEntry.size;
+            item->_modtime = localEntry.modtime;
+            item->_type = localEntry.isDirectory ? ItemTypeDirectory : ItemTypeFile;
+            if (!_dirItem || _dirItem->_direction == SyncFileItem::Down) {
+                _childModified = true;
+            }
+        } else if (!dbEntry.isValid()) { // New local file
+            item->_instruction = CSYNC_INSTRUCTION_NEW;
+            item->_direction = SyncFileItem::Up;
+            item->_checksumHeader.clear();
+            item->_size = localEntry.size;
+            item->_modtime = localEntry.modtime;
+            item->_type = localEntry.isDirectory ? ItemTypeDirectory : localEntry.isVirtualFile ? ItemTypeVirtualFile : ItemTypeFile;
+            if (!_dirItem || _dirItem->_direction == SyncFileItem::Down) {
+                _childModified = true;
+            }
+
+            auto postProcessLocalNew = [item, localEntry, this]() {
+                if (localEntry.isVirtualFile) {
+                    // Remove the spurious file if it looks like a placeholder file
+                    // (we know placeholder files contain " ")
+                    if (localEntry.size <= 1) {
+                        qCWarning(lcDisco) << "Wiping virtual file without db entry for" << _currentFolder._local + "/" + localEntry.name;
+                        item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                        item->_direction = SyncFileItem::Down;
+                    } else {
+                        qCWarning(lcDisco) << "Virtual file without db entry for" << _currentFolder._local << localEntry.name
+                                           << "but looks odd, keeping";
+                        item->_instruction = CSYNC_INSTRUCTION_IGNORE;
+                    }
+                }
+            };
+
+            // Check if it is a move
+            OCC::SyncJournalFileRecord base;
+            if (!_discoveryData->_statedb->getFileRecordByInode(localEntry.inode, &base)) {
+                dbError();
```

Comment:
This emits `finished()`, but `scheduleSubJobs()` will still be called via a timer from `process()` leading to a double invocation of `finished()`. Something similar happens for the other rename detection.

Reply:
well spotted.

dbError is not tested at all :-(

Maybe it make sense to make a test for it...  (where we remove the db or cause it to error somehow)